MonumentAddons.cs

```diff
@@ -3631,6 +3634,205 @@ private interface IDeepCollection
             return false;
         }
 
+        private class DungeonEntranceMapper
+        {
+            private Dictionary<DungeonGridInfo, MonumentInfo> _entranceToMonument = new Dictionary<DungeonGridInfo, MonumentInfo>();
+
+            public void OnServerInitialized()
+            {
+                foreach (var monumentInfo in TerrainMeta.Path.Monuments)
+                {
+                    if (monumentInfo.DungeonEntrance != null)
+                    {
+                        _entranceToMonument[monumentInfo.DungeonEntrance] = monumentInfo;
+                    }
+                }
+            }
+
+            public MonumentInfo GetMonumentFromTunnel(DungeonGridCell dungeonGridCell)
+            {
+                var entrance = TerrainMeta.Path.FindClosest(TerrainMeta.Path.DungeonGridEntrances, dungeonGridCell.transform.position);
+                if (entrance == null)
+                    return null;
+
+                return GetMonumentFromEntrance(entrance);
+            }
+
+            private MonumentInfo GetMonumentFromEntrance(DungeonGridInfo dungeonGridInfo)
+            {
+                MonumentInfo monumentInfo;
+                return _entranceToMonument.TryGetValue(dungeonGridInfo, out monumentInfo)
+                    ? monumentInfo
+                    : null;
+            }
+        }
+
+        private static class PhoneUtils
+        {
+            public const string Tunnel = "FTL";
+            public const string UnderwaterLab = "Underwater Lab";
+            public const string CargoShip = "Cargo Ship";
+
+            private static readonly Regex SplitCamelCaseRegex = new Regex("([a-z])([A-Z])", RegexOptions.Compiled);
+
+            public static void NameTelephone(Telephone telephone, SingleEntityAdapter adapter, MonumentAddons plugin) 
```

Comment:
For now I removed adapter from helper parameters. Where I should instantiate `MonumentHelper` class? I was thinking to put it where `DungeonEntranceMapper ` currently is (plugin property), and then pass it to `PhoneUtils` in `SingleEntityAdapter` but it looks to me like almost same thing. Alternatively new instance for every `PhoneUtils.NameTelephone` call, but it seems to be

```C#
var telephone = Entity as Telephone;
if (telephone != null && telephone.prefabID == 1009655496)
{
    PhoneUtils.NameTelephone(telephone, Monument, Position, PluginInstance._monumentHelper);
}
```

Reply:
`MonumentHelper` instance should be a field of the plugin, similar to the `DungeonEntranceMapper` instance. Could even merge the `DungeonEntranceMapper` code into the `MonumentHelper` class to reduce the number of different helpers needed. Definitely don't create more instances since we want to avoid heap allocations that will be quickly dereferenced (a.k.a. "garbage"). Above looks fine.

Reply:
`DungeonEntranceMapper` merged into `MonumentHelper`, removed plugin instance from `Phone utils`